source/RoslynAnalyzers/PossibleUnintentionalCreationOfEnumeratorAnalyzer.cs

```diff
@@ -0,0 +1,170 @@
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Octopus.RoslynAnalyzers
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class PossibleUnintentionalCreationOfEnumeratorAnalyzer : DiagnosticAnalyzer
+    {
+        const string DiagnosticId = "Octopus_PossibleUnintentionalCreationOfEnumerator";
+
+        const string Title = "Any(), Count() or None() call likely unintentionally creates an enumerator";
+
+        const string MessageFormat = "This Any(), Count() or None() call on this type likely creates an enumerator unintentionally. This can cause performance problems. Use our custom extension methods or the Count or Length properties instead.";
+        const string Category = "Octopus";
+
+        const string Description = @"The Any() and Count() extension methods cause the target to be enumerated unless it 
+implements IListProvider<T>, ICollection<T> or ICollection. Most collections do this, but some only implement 
+IEnumerable<T>. The creation of an enumerator is expensive relative to calling the Count or Length properties, 
+so it should be avoided when possible. We have several implementations of a None() extension method which calls 
+Any(). A best effort is used to catch this usage as they should mainly be under the Octopus namespace.";
+
+        internal static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(
+            DiagnosticId,
+            Title,
+            MessageFormat,
+            Category,
+            DiagnosticSeverity.Error,
+            true,
+            Description
+        );
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+            context.RegisterSyntaxNodeAction(CheckForUnexpectedEnumeration, SyntaxKind.InvocationExpression);
+        }
+
+        void CheckForUnexpectedEnumeration(SyntaxNodeAnalysisContext context)
+        {
+            if (
+                context.Node is InvocationExpressionSyntax invocation &&
```

Comment:
You can just cast this directly as you already registered for just `SyntaxKind.InvocationExpression` nodes.